File: LeetCode/Python/solutions/p0017.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        mp = [
            [], # index 0 garbage
            [], # number 1 maps to nothing
            ["a", "b", "c"],
            ["d", "e", "f"],
            ["g", "h", "i"],
            ["j", "k", "l"],
            ["m", "n", "o"],
            ["p", "q", "r", "s"],
            ["t", "u", "v"],
            ["w", "x", "y", "z"]
        ]

        #
        # Solution v1: Brute Force
        #
        # Runtime: 29 ms @ (beats) 90.95%
        # Memory Usage: 13.9 MB @ (beats) 27.70%
        #

        def func(resp, digit):
            if resp:
                temp = []
                for s in resp:
                    for c in mp[digit]:
                        temp.append(f"{s}{c}")
                    print(temp)

                return temp
            else:
                return mp[digit]

        length = len(digits)
        if length == 0:
            return []
        elif length == 1:
            return mp[int(digits[0])]
        else:
            ans = []
            for d in digits:
                print("else")
                ans = func(ans, int(d))

            return ans


if __name__ == "__main__":
    solution = Solution()

    digits = "23"
    print(f"return: {solution.letterCombinations(digits)}")

    digits = ""
    print(f"return: {solution.letterCombinations(digits)}")

File: LeetCode/Java/src/solutions/p0017_LetterCombinationsPhoneNumber/P0017.java
package solutions.p0017_LetterCombinationsPhoneNumber;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class P0017 {

    public List<String> letterCombinations(String digits) {
        /*
         * Solution v1: Brute Force
         *
         * Runtime: 1 ms @ (beats) 79.70%
         * Memory Usage: 41.1 MB @ (beats) 69.15%
         */
        List<String> ans = new ArrayList<>();
        if (digits == null || digits.length() == 0) {
            return ans;
        }

        for (int i = 0; i < digits.length(); i++) {
            ans = buider(new ArrayList<String>(), (int) (digits.charAt(i)));
        }

        return ans;
    }

    public static List<String> buider(List<String> ans, int d) {
        String[][] mp = new String[][]{
                {},
                {},
                {"a", "b", "c"},
                {"d", "e", "f"},
                {"g", "h", "i"},
                {"j", "k", "l"},
                {"m", "n", "o"},
                {"p", "q", "r", "s"},
                {"t", "u", "v"},
                {"w", "x", "y", "z"}
        };

        if (ans == null || ans.size() == 0) {
            return Arrays.asList(mp[d]);
        } else {
            List<String> temp = new ArrayList<>();

            for (String a : ans) {
                for (String s : mp[d]) {
                    temp.add(a.concat(s));
                }
            }
            return temp;
        }
    }
}
